src/wallet/did_wallet/did_wallet.py

```diff
@@ -0,0 +1,760 @@
+        removal_amount = 0
+
+        for record in unconfirmed_tx:
+            if record.incoming:
+                addition_amount += record.amount
+            else:
+                removal_amount += record.amount
+
+        result = confirmed - removal_amount + addition_amount
+
+        self.log.info(f"Unconfirmed balance for did wallet is {result}")
+        return uint64(result)
+
+    async def select_coins(
+        self, amount, exclude: List[Coin] = None
+    ) -> Optional[Set[Coin]]:
+        """ Returns a set of coins that can be used for generating a new transaction. """
+        async with self.wallet_state_manager.lock:
+            if exclude is None:
+                exclude = []
+
+            spendable_amount = await self.get_spendable_balance()
+
+            if amount > spendable_amount:
+                self.log.warning(
+                    f"Can't select amount higher than our spendable balance {amount}, spendable {spendable_amount}"
+                )
+                return None
+
+            self.log.info(f"About to select coins for amount {amount}")
+            unspent: List[WalletCoinRecord] = list(
+                await self.wallet_state_manager.get_spendable_coins_for_wallet(
+                    self.wallet_info.id
+                )
+            )
+            sum_value = 0
+            used_coins: Set = set()
+
+            # Use older coins first
+            unspent.sort(key=lambda r: r.confirmed_block_index)
+
+            # Try to use coins from the store, if there isn't enough of "unused"
+            # coins use change coins that are not confirmed yet
+            unconfirmed_removals: Dict[
+                bytes32, Coin
+            ] = await self.wallet_state_manager.unconfirmed_removals_for_wallet(
+                self.wallet_info.id
+            )
+            for coinrecord in unspent:
+                if sum_value >= amount and len(used_coins) > 0:
+                    break
+                if coinrecord.coin.name() in unconfirmed_removals:
+                    continue
+                if coinrecord.coin in exclude:
+                    continue
+                sum_value += coinrecord.coin.amount
+                used_coins.add(coinrecord.coin)
+                self.log.info(
+                    f"Selected coin: {coinrecord.coin.name()} at height {coinrecord.confirmed_block_index}!"
+                )
+
+            # This happens when we couldn't use one of the coins because it's already used
+            # but unconfirmed, and we are waiting for the change. (unconfirmed_additions)
+            if sum_value < amount:
+                raise ValueError(
+                    "Can't make this transaction at the moment. Waiting for the change from the previous transaction."
+                )
+
+        self.log.info(f"Successfully selected coins: {used_coins}")
+        return used_coins
+
+    # This will be used in the recovery case where we don't have the parent info already
```

Comment:
I don't understand this comment.

Reply:
You can recover a DID. So most of the time you will already have you parent info stored because you are just sending the coin back to yourself usually, however if you have a DID Wallet which is new and doesn't have an associated DID yet you need to learn the parent info by inspecting the block which is what this function does